core/lib/spree/event/event.rb

```diff
@@ -18,9 +18,29 @@ class Event
       # @return [Hash]
       attr_reader :payload
 
+      # Creation time for the {Event} instance
+      #
+      # @return [Time]
+      attr_reader :time
+
+      # Relevant stack location for the creation of the {Event}
+      #
+      # It's set on initialization. It defaults to 4 levels up the stack to
+      # go through:
+      #
+      #   - The {#initialize} method.
+      #   - The adapter `#fire` method (for instance
+      #   {Spree::Events::Adapters#fire}).
+      #   - {Spree::Event.fire}.
+      #
+      # @return [Thread::Backtrace::Location]
+      attr_reader :caller_location
+
       # @api private
-      def initialize(payload:)
+      def initialize(payload: {}.freeze, caller_location: caller_locations(4, 1)[0])
         @payload = payload
+        @time = Time.now
```

Comment:
Hmmm, good point. Thinking about it twice, definitely, we shouldn't use `Time.now`, as the TZ that it'll use doesn't need to match the Rails one.

However, `Time.current` is an `ActiveSupport` method. It's not a big deal, but the event bus sub-system (besides the legacy adapter) is a pure Ruby implementation. I.e., it could be extracted as a gem at some point, and it'd work with any ruby project. What do you think if we use `UTC` or Unix time instead? We can be explicit with the naming and call it `utc_time` or `unix_time`. Using Unix time could help in the future if we find a way to support async subscribers, but of course, we'll have to consider much more than only that time attribute.

Reply:
I've read a lot of people struggle with issues related to not using timezones everywhere in Rails applications, I myself had a few issues in the past when comparing times transformed into dates or using `Time.now` in tests, so I'm possibly overcautious here. But I think that using `UTC` and being explicit in the naming is a good idea, so I'll be OK with it 👍